### DISCUSSION TITLE
Performance Optimization for Large Number of Portal Users (120,000+)

### DISCUSSION THREAD

**Discussion by Abhijithsoman** [OCA/odoo-community.org#170] [Q&A] [answered]
We're facing performance issues in our Odoo installation, particularly with list views and user management tasks, when dealing with a large number of portal users (over 120,000 records in the res_users table).Its takes huge amount of time to  load the list view more than 5 hours.
Even executing a delete query directly in the database takes minutes unless we disable the triggers.This issue seems to be only in the res_users table. Also the system loads continuously if we try to edit a user record but the issue is only if we try to modify the login field. We've indexed relevant fields in the res_users table, but haven't observed any improvements.

Are there any other strategies or best practices we should consider for handling a large number of portal users.?

note - The portal users here are not interacting with the web application directly they are created from an external API and used for recording 
 the user information in the back-end system . But the internal users need to access the views and need to perform modifications on  internal user records (Create ,Write or Delete Internal users)

**Comment by rvalyi** [comment, marked as answer]
it would help if you specificify the Odoo version you are using. Also if you are able to list the installed modules that could possibly extend the portal features or contacts or objects you are showing in the portal that would help too... Cause Odoo is usually expected to deal well with large volume of records.

Iideally you should start by looking what is making the CPUs busy (using top): Postgresql or python mostly. If it's mostly Postgresql, use approaches like this https://simplebackups.com/blog/postgresql-how-to-find-long-running-queries/ to hunt for slow queries and the   track their origin in the Odoo code (then possibly add indexes, avoid them etc). Also hunt for repeating SQL queries and track the possible loops/"N+1 queries" issues in Odoo. These are the low hanging fruits. If it doesn't fix the issue you should then dig into the craziness of the computed fields chains. But yeah in any case optimizing 5 hours requests won't be easy. You might also consider hiring the big guys in the OCA to get it fixed.

**Comment by yajo** [comment]
The bad part about these kind of performance problems is that they are usually only discovered once they happen, and you need to use your environment, where it's happening, to diagnose and fix the issue. Each performance issue is different, but it generally narrows down to:
- Non-optimized SQL queries. See comment above and https://www.youtube.com/live/DmRei44Ka0I?feature=shared
- Non-optimized Python code. See https://www.youtube.com/live/5jT5eoCZH6U?feature=shared and https://github.com/benfred/py-spy
- Hardware malfunction. Make sure CPU has proper cooling, and that you have enough RAM. Also check how full is the disk. Basic things, but they affect performance.

**Comment by antoniospneto** [comment]
This here also helps a lot to discover the source of bad code:
https://www.odoo.com/documentation/14.0/developer/howtos/profilecode.html

**Comment by Abhijithsoman** [comment]
We are using odoo13 community edition and the issue is almost fixed temporarily. There are two issues which we  were facing one is loading time of list view when we remove the internal users filter from the list view and the another one is huge time for doing create /write in the res_users table.

And for the first one  when we do the analysis of the query below which is running 

SELECT "res_users".id FROM "res_users","res_partner" as "res_users__partner_id" WHERE ("res_users"."active" = true) AND (("res_users"."share" IS NULL or "res_users"."share" = false ) OR ("res_users"."id" in (SELECT "user_id" FROM "res_company_users_rel" WHERE "cid" IN (1, 4)))) AND ("res_users"."partner_id" = "res_users__partner_id"."id") AND ((("res_users__partner_id"."partner_share" IS NULL or "res_users__partner_id"."partner_share" = false ) OR ("res_users__partner_id"."company_id" in (1,4))) OR "res_users__partner_id"."company_id" IS NULL ) ORDER BY "res_users__partner_id"."name" ,"res_users"."login"

("res_users"."share" IS NULL or "res_users"."share" = false )  This condition was taking most of the time for loading.Which is the default internal users filter.

Also the reason for huge time on create/write in the records was due to the huge size of mail_message table it is fixed by deleting the records.

Another issue was the creation of users with duplicate logins seems like the unique constraints on login is not working. may be related to below issues

https://postgresql.verite.pro/blog/2018/08/27/glibc-upgrade.html

https://dba.stackexchange.com/questions/268027/postgresql-duplicates-despite-having-a-unique-constraint
